File: Sources/AppFamille/Models/Recurrence.swift
import Foundation

/// Règle de récurrence simple pour le MVP. On couvre les cas les plus fréquents
/// d'une famille (poubelles chaque semaine, loyer chaque mois...) sans viser
/// la complexité complète de RFC 5545 — on pourra l'étendre plus tard si besoin.
enum RecurrenceRule: String, Codable, CaseIterable, Identifiable {
    case none
    case daily
    case weekly
    case monthly
    case yearly

    var id: String { rawValue }

    var label: String {
        switch self {
        case .none: return "Ne se répète pas"
        case .daily: return "Tous les jours"
        case .weekly: return "Toutes les semaines"
        case .monthly: return "Tous les mois"
        case .yearly: return "Tous les ans"
        }
    }

    /// Calcule la prochaine occurrence après `date`, ou nil si la règle est `.none`.
    func nextOccurrence(after date: Date, calendar: Calendar = .current) -> Date? {
        switch self {
        case .none:
            return nil
        case .daily:
            return calendar.date(byAdding: .day, value: 1, to: date)
        case .weekly:
            return calendar.date(byAdding: .weekOfYear, value: 1, to: date)
        case .monthly:
            return calendar.date(byAdding: .month, value: 1, to: date)
        case .yearly:
            return calendar.date(byAdding: .year, value: 1, to: date)
        }
    }
}

File: web/src/models/recurrence.ts
/**
 * Règle de récurrence simple pour le MVP : on couvre les cas les plus
 * fréquents d'une famille (poubelles chaque semaine, loyer chaque mois...)
 * sans viser la complexité complète d'iCalendar (RFC 5545) — on pourra
 * l'étendre plus tard si besoin.
 */
export type RecurrenceRule = 'none' | 'daily' | 'weekly' | 'monthly' | 'yearly'

export const RECURRENCE_LABELS: Record<RecurrenceRule, string> = {
  none: 'Ne se répète pas',
  daily: 'Tous les jours',
  weekly: 'Toutes les semaines',
  monthly: 'Tous les mois',
  yearly: 'Tous les ans',
}

export const RECURRENCE_OPTIONS: RecurrenceRule[] = ['none', 'daily', 'weekly', 'monthly', 'yearly']

/** Calcule la prochaine occurrence après `date`, ou null si la règle est 'none'. */
export function nextOccurrence(rule: RecurrenceRule, date: Date): Date | null {
  const next = new Date(date)
  switch (rule) {
    case 'none':
      return null
    case 'daily':
      next.setDate(next.getDate() + 1)
      return next
    case 'weekly':
      next.setDate(next.getDate() + 7)
      return next
    case 'monthly':
      next.setMonth(next.getMonth() + 1)
      return next
    case 'yearly':
      next.setFullYear(next.getFullYear() + 1)
      return next
  }
}
